File: app/src/main/java/com/fenfei/myapplicationdemo/views/MyTextview.java
package com.fenfei.myapplicationdemo.views;

import android.content.Context;
import android.support.v4.view.MotionEventCompat;
import android.support.v7.widget.CardView;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.widget.Button;

/**
 * Created by shefenfei on 2017/1/23.
 */
public class MyTextview extends Button {

    private String TAG = this.getClass().getSimpleName();

    public MyTextview(Context context) {
        this(context, null);
    }

    public MyTextview(Context context, AttributeSet attrs) {
        this(context, attrs , 0);
    }

    public MyTextview(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        CardView cardView;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        boolean intercept = false;
        int action = MotionEventCompat.getActionMasked(event);
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                Log.e(TAG, "onTouchEvent: " + "ACTION_DOWN");
                intercept = true;
                return intercept;
//                break;

            case MotionEvent.ACTION_MOVE:
                Log.e(TAG, "onTouchEvent: " + "ACTION_MOVE");
                intercept = false;
                break;
            case MotionEvent.ACTION_UP:
                Log.e(TAG, "onTouchEvent: " + "ACTION_UP");
                break;
            case MotionEvent.ACTION_CANCEL:
                Log.e(TAG, "onTouchEvent: " + "ACTION_CANCEL");
        }
        return super.onTouchEvent(event);
//        return intercept;
    }

    /*
    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        int action = event.getAction();
        switch (action)
        {
            case MotionEvent.ACTION_DOWN:
                Log.e(TAG, "dispatchTouchEvent ACTION_DOWN");
//                return true;
                break;
            case MotionEvent.ACTION_MOVE:
                Log.e(TAG, "dispatchTouchEvent ACTION_MOVE");
                break;
            case MotionEvent.ACTION_UP:
                Log.e(TAG, "dispatchTouchEvent ACTION_UP");
                break;

            default:
                break;
        }
        return super.dispatchTouchEvent(event);    //事件会自动的分发给当前 View 的 onInterceptTouchEvent 方法 在自定义view中并不会有这个方法
//        return true; //直接返回true 代表我将直接对事件消费,并不会向下传递,不会到下面的onTouchEvent();
    }

    */




}

File: camera/src/main/java/com/fenfei/camera/kotlins/views/MyTextView.kt
package com.fenfei.camera.kotlins.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.animation.Animation
import android.widget.TextView

/**
 * Created by shefenfei on 2017/6/14.
 */
class MyTextView(override val count: Int, context: Context?, attrs: AttributeSet?) : TextView(context, attrs), Foo {
    override fun f() {

    }

    lateinit var paint: Paint

    init {
        initView()
    }

    fun initView() {
        paint = Paint()
        paint.color = Color.RED
        paint.isAntiAlias = true
    }

    override fun getAnimation(): Animation {
        return super.getAnimation()
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        canvas?.drawText("hahahhaha", 100f, 100f, paint)
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
//        return super.onTouchEvent(event)
        return true
    }

}


interface Foo {
    val count: Int

    fun f()
}


class Bar1 : Foo , Bar3() {
    override val count: Int
        get() = 1


    override fun f() {
//        super<Foo>.f()
        super<Bar3>.f()
    }
}

class Bar2 : Foo {
    override fun f() {

    }

    override val count: Int
        get() = 1
}


open class Bar3 : Impl() {
    override fun f() {

    }

}

open class Base {
    open fun f() {}
}

abstract class Impl : Base() {
    override abstract fun f()
}

//在 Kotlin 中类没有静态方法 在大多数情况下，它建议简单地使用包级函数。

class Address {
    var name: String = ""
    var age: Int = 1
}

fun copyAddress(address: Address) : Address {
    val result = Address()
    result.age = address.age
    result.name = address.name
    return result
}


val username: String = ""
